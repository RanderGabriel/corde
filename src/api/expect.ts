import { expectMatchWithNot, MatchesWithNot } from "./matcher";

/**
 * Receives wich command will be tested.
 *
 * @param commandName Command name.
 *
 * @description Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @returns The **Compare** object, where will handle
 * the type of response is expected.
 */
export function expect(commandName: string): MatchesWithNot {
  return expectMatchWithNot(commandName);
}
