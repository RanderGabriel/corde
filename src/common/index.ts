export * from './cordeBot';
export * from './runtime';
export * from './testColletor';
