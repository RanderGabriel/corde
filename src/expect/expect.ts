import { VoidLikeFunction } from "../types";
import { ExpectMatchesWithNot, MatchWithNot } from "./matcher";

/**
 * Invalid declaration of a test matcher.
 *
 * @since 3.0
 */
export function expect(): void;
/**
 * Receives which command will be tested.
 *
 * Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @param commandNameResolvable Command name. (Empty strings will resolve failed test)
 *
 * @returns An object with all possible tests to be done
 * in the bot.
 *
 * @since 1.0
 */
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable: T,
): MatchWithNot;
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable?: T,
): MatchWithNot | void {
  if (commandNameResolvable == null) {
    return;
  }
  return new ExpectMatchesWithNot(commandNameResolvable);
}
