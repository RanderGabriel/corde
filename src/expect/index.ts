import { RoleMatches, MessageMatches } from "./matcher";
import { Expect, MacherContructorArgs } from "./types";

function getMessageMatchers(): string[] {
  return getFunctions(MessageMatches);
}

function getRoleMatchers(): string[] {
  return getFunctions(RoleMatches);
}

function getFunctions<T>(type: new (...args: any[]) => T): string[] {
  return Object.getOwnPropertyNames(type.prototype).filter(
    (propName) => propName !== "constructor" && typeof type.prototype[propName] === "function",
  );
}

const expectation = {
  not: {},
  inChannel: {},
  inGuild: {},
};

function createTestFunction(classType: any, params: MacherContructorArgs, functionName: string) {
  return (...args: any[]) => (new classType(params) as any)[functionName](...args);
}

function createTestsFromMatches<T>(
  names: string[],
  params: MacherContructorArgs,
  classType: new (...args: any[]) => T,
): [string, (...args: any[]) => any][] {
  return names.map((name) => {
    return [name, createTestFunction(classType, params, name)];
  });
}

function set(from: [string, (...args: any[]) => any][], to: any) {
  from.forEach((test) => {
    to[test[0]] = test[1];
  });
}

const messageTestNames = getMessageMatchers();
const roleMatchers = getRoleMatchers();

const _expect: any = <T extends (() => number | string) | number | string>(commandName: T) => {
  const messageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: false },
    MessageMatches,
  );

  const isNotMessageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: true },
    MessageMatches,
  );

  const roleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: false },
    RoleMatches,
  );

  const isNotRoleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: true },
    RoleMatches,
  );

  set(messageTests, expectation);
  set(isNotMessageTests, expectation.not);

  set(roleTests, expectation);
  set(isNotRoleTests, expectation.not);

  expectation.inChannel = (channelId: string) => {
    const channelMatchers: any = {
      not: {},
    };

    const messageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId },
      MessageMatches,
    );

    const isNotmessageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId, isNot: true },
      MessageMatches,
    );

    set(messageTests, channelMatchers);
    set(isNotmessageTests, channelMatchers.not);

    return channelMatchers;
  };

  expectation.inGuild = (guildId: string) => {
    const guildMatchers: any = {
      not: {},
    };

    const roleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    const isNotRoleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    set(roleTests, guildMatchers);
    set(isNotRoleTests, guildMatchers.not);

    return guildMatchers;
  };

  return expectation;
};

const messageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: false, isCascade: true },
  MessageMatches,
);

const isNotMessageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: true, isCascade: true },
  MessageMatches,
);

const roleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: false, isCascade: true },
  RoleMatches,
);

const isNotRoleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: true, isCascade: true },
  RoleMatches,
);

_expect.not = {};

set(messageTests, _expect);
set(isNotMessageTests, _expect.not);

set(roleTests, _expect);
set(isNotRoleTests, _expect.not);

const expect = _expect as Expect;

export { expect };
