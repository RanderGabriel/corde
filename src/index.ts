/**
 * @file Automatically generated by barrelsby.
 */

 export * from "./cli";
export * from "./erros/index";
