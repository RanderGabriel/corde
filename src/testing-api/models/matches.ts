import { MessageEmbed } from 'discord.js';

/**
 * Defines all functions that can be used
 * to check a bot reaction of a command.
 */
export interface Matches {
  /**
   * Defines the message expected to be returned by a
   * command.
   *
   * @param expect A message returned by a bot after invoke a command
   *
   */
  mustReturn(expect: string | MessageEmbed): void;
  /**
   * Defines reactions that must be add to command message.
   *
   * @param reaction Single or list of reactions that must be added to an message
   *
   * @see For how to react message -> https://discordjs.guide/popular-topics/reactions.html#reacting-to-messages
   *
   * @example
   *
   *  bot.on('message', async (message) => {
   *    if (command === 'emoji') {
   *       msg.react('😄');
   *    } else if(command === 'emojis') {
   *       Promise.all([msg.react('😄'), msg.react('🍊')]);
   *    }
   *  });
   *
   *  Tests:
   *
   *  command('emoji').mustAddReaction('😄');
   *  command('emojis').mustAddReaction('😄', '🍊');
   */
  mustAddReaction(...reaction: string[]): void;
}
