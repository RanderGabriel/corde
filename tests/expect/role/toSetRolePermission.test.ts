import { Client } from "discord.js";
import { ToSetRolePermission } from "../../../src/expect/matches";
import MockDiscord from "../../mocks/mockDiscord";
import { initCordeClientWithChannel } from "../../testHelper";
import { TestReport } from "../../../src/types";
import { buildReportMessage, calcPermissionsValue, Permission } from "../../../src/utils";
import { MockEvents } from "../../mocks/mockEvents";
import { runtime } from "../../../src/common";

let mockDiscord = new MockDiscord();

function initClient() {
  const corde = initCordeClientWithChannel(mockDiscord, new Client());
  corde.findRole = jest.fn().mockReturnValue(mockDiscord.role);
  corde.fetchRole = jest.fn().mockReturnValue(mockDiscord.role);
  corde.sendTextMessage = jest.fn().mockImplementation(() => {});
  return corde;
}

describe("testing toSetRolePermission operation", () => {
  afterEach(() => {
    mockDiscord = new MockDiscord();
  });

  it("should fail due to undefined roleIdentifier", async () => {
    const corde = initClient();
    const toSetRolePermission = new ToSetRolePermission(corde, "test", false);
    const report = await toSetRolePermission.action(undefined, ["ADD_REACTIONS"]);

    const message = buildReportMessage(
      "expected: data to identifier the role (id or name)\n",
      `received: null`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to invalid permission parameter (object)", async () => {
    const corde = initClient();
    const toSetRolePermission = new ToSetRolePermission(corde, "test", false);
    // @ts-ignore
    const report = await toSetRolePermission.action({ id: "123" }, {});

    const message = buildReportMessage(
      `expected: permissions to be null, undefined or an array\n`,
      `received: object`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to invalid permission parameter (undefined)", async () => {
    const corde = initClient();
    const toSetRolePermission = new ToSetRolePermission(corde, "test", false);
    // @ts-ignore
    const report = await toSetRolePermission.action({ id: "123" }, {});

    const message = buildReportMessage(
      `expected: permissions to be null, undefined or an array\n`,
      `received: object`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to not found role", async () => {
    const corde = initClient();
    corde.findRole = jest.fn().mockReturnValue(null);
    const toSetRolePermission = new ToSetRolePermission(corde, "test", false);
    const report = await toSetRolePermission.action({ id: "123" }, ["ATTACH_FILES"]);

    const message = buildReportMessage(`expected: role with id 123\n`, `received: null`);

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to role permissions was not updated", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);

    const toSetRolePermission = new ToSetRolePermission(corde, "test", false);
    const report = await toSetRolePermission.action({ id: "123" }, ["ATTACH_FILES"]);

    const message = buildReportMessage(
      `expected: role permissions change to: ATTACH_FILES\n`,
      `received: permissions were not changed`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return true due to isNot true and no role change", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);

    const toSetRolePermission = new ToSetRolePermission(corde, "test", true);
    const report = await toSetRolePermission.action({ id: "123" }, ["ATTACH_FILES"]);

    const expectReport: TestReport = {
      pass: true,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return true due to role changed permissions (isNot false)", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);
    const mockRole = mockDiscord.createMockRole(
      "test role",
      calcPermissionsValue(...mockDiscord.role.permissions.toArray().map((p) => Permission[p])),
    );
    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const toSetRolePermission = new ToSetRolePermission(corde, "test", false);
    const report = await toSetRolePermission.action({ id: "123" }, mockRole.permissions.toArray());

    const expectReport: TestReport = {
      pass: true,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due to permissions should not change (isNot true)", async () => {
    const corde = initClient();
    const mockRole = mockDiscord.createMockRole(
      "test role",
      calcPermissionsValue(Permission.ADMINISTRATOR),
    );
    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const toSetRolePermission = new ToSetRolePermission(corde, "test", true);
    const report = await toSetRolePermission.action({ id: "123" }, ["ADMINISTRATOR"]);

    const message = buildReportMessage(
      `expected: role permissions not change to: ADMINISTRATOR\n`,
      `received: ADMINISTRATOR (and ${
        mockRole.permissions.toArray().filter((p) => p !== "ADMINISTRATOR").length
      } others)`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due to permissions should not change (isNot true)", async () => {
    const corde = initClient();
    const mockRole = mockDiscord.createMockRole(
      "test role",
      calcPermissionsValue(Permission.ADMINISTRATOR, Permission.BAN_MEMBERS),
    );
    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const toSetRolePermission = new ToSetRolePermission(corde, "test", true);
    const report = await toSetRolePermission.action({ id: "123" }, [
      "ADMINISTRATOR",
      "BAN_MEMBERS",
    ]);

    const message = buildReportMessage(
      `expected: role permissions not change to: ADMINISTRATOR and BAN_MEMBERS\n`,
      `received: ADMINISTRATOR (and ${
        mockRole.permissions.toArray().filter((p) => p !== "ADMINISTRATOR").length
      } others)`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due to permissions should not change (isNot true)", async () => {
    const corde = initClient();
    const mockRole = mockDiscord.createMockRole(
      "test role",
      calcPermissionsValue(Permission.ADMINISTRATOR, Permission.BAN_MEMBERS, Permission.CONNECT),
    );
    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const toSetRolePermission = new ToSetRolePermission(corde, "test", true);
    const report = await toSetRolePermission.action({ id: "123" }, [
      "ADMINISTRATOR",
      "BAN_MEMBERS",
      "CONNECT",
    ]);

    const message = buildReportMessage(
      `expected: role permissions not change to: ADMINISTRATOR (and 2 others)\n`,
      `received: ADMINISTRATOR (and ${
        mockRole.permissions.toArray().filter((p) => p !== "ADMINISTRATOR").length
      } others)`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due expected name did not match to received", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);
    mockEvent.mockOnceRolePermissionsUpdate(mockDiscord.role);
    const toSetRolePermission = new ToSetRolePermission(corde, "test", false);
    const report = await toSetRolePermission.action({ id: "123" }, ["ATTACH_FILES"]);

    const message = buildReportMessage(
      `expected: role permissions change to: ATTACH_FILES\n`,
      `received: ADMINISTRATOR (and 30 others)`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });
});
