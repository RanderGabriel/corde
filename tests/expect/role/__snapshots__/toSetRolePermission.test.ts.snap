// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing toSetRolePermission operation should fail due to no role mentionable was not updated 1`] = `
Object {
  "message": "    
    expected: role permissions change to: ATTACH_FILES
    received: permissions were not changed
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should fail due to undefined roleIdentifier 1`] = `
Object {
  "message": "    
    expected: data to identifier the role (id or name)
    received: null
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should return a not passed test due expected name did not match to received 1`] = `
Object {
  "message": "    
    expected: role permissions change to: ATTACH_FILES
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should return a not passed test due to mentionable should not change (isNot true) 1`] = `
Object {
  "message": "    
    expected: role permissions not change to: ADMINISTRATOR
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should return a not passed test due to mentionable should not change (isNot true) 2`] = `
Object {
  "message": "    
    expected: role permissions not change to: ADMINISTRATOR and BAN_MEMBERS
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should return a not passed test due to mentionable should not change (isNot true) 3`] = `
Object {
  "message": "    
    expected: role permissions not change to: ADMINISTRATOR (and 2 others)
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should return false due to invalid mentionable parameter (object) 1`] = `
Object {
  "message": "    
    expected: permissions to be null, undefined or an array
    received: object
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should return false due to invalid mentionable parameter (undefined) 1`] = `
Object {
  "message": "    
    expected: permissions to be null, undefined or an array
    received: object
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should return false due to not found role 1`] = `
Object {
  "message": "    
    expected: role with id 123
    received: null
",
  "pass": false,
}
`;

exports[`testing toSetRolePermission operation should return true due to isNot true and no role change 1`] = `
Object {
  "pass": true,
}
`;

exports[`testing toSetRolePermission operation should return true due to role changed the mentionable (isNot false) 1`] = `
Object {
  "pass": true,
}
`;
